src/render.rs

```diff
@@ -0,0 +1,236 @@
+    fn sender(&self) -> &UserId;
+}
+
+impl RenderableEvent for EncryptedEvent {
+    // TODO: this is not implemented yet
+    fn render(&self, displayname: &str) -> String {
+        let color_user = Weechat::color("green"); // TODO: get per-user color
+        let color_reset = Weechat::color("reset");
+        format!(
+            "{color_user}{}{color_reset}\t{}",
+            displayname,
+            "Unable to decrypt message",
+            color_user = color_user,
+            color_reset = color_reset
+        )
+    }
+
+    fn sender(&self) -> &UserId {
+        &self.sender
+    }
+}
+
+impl RenderableEvent for MemberEvent {
+    fn render(&self, displayname: &str) -> String {
+        use MembershipChange::*;
+        let change_op = self.membership_change();
+        let operation = match change_op {
+            None => "did nothing",
+            Error => "caused an error", // must never happen
+            Joined => "has joined the room",
+            Left => "has left the room",
+            Banned => "was banned",
+            Unbanned => "was unbanned",
+            Kicked => "was kicked from the room",
+            Invited => "was invited to the room",
+            KickedAndBanned => "was kicked and banned",
+            InvitationRejected => "rejected the invitation",
+            InvitationRevoked => "had the invitation revoked",
+            ProfileChanged => "changed the display name or avatar",
+            NotImplemented => "performed an unimplemented operation",
+        };
+
+        let (prefix, color_action) = match change_op {
+            Joined | Invited => ("join", "green"),
+            _ => ("quit", "red"),
+        };
+
+        format!(
+            "{prefix}{color_user}{name} {color_deli}({color_reset}{state}{color_deli}){color_reset} {color_action}{op}",
+            prefix = Weechat::prefix(prefix),
+            color_user = Weechat::color("reset"), // TODO
+            color_deli = Weechat::color("chat_delimiters"),
+            color_action = Weechat::color(color_action),
+            color_reset = Weechat::color("reset"),
+            name = displayname,
+            state = self.state_key,
+            op = operation,
+        )
+    }
+
+    fn sender(&self) -> &UserId {
+        &self.sender
+    }
+}
+
+impl RenderableEvent for MessageEvent {
+    fn render(&self, displayname: &str) -> String {
+        use MessageEventContent::*;
+
+        let color_user = Weechat::color("green"); // TODO: get per-user color
+        let color_reset = Weechat::color("reset");
+        match &self.content {
+            // TODO: formatting for inline markdown and so on
+            Text(t) => format!(
+                "{color_user}{}{color_reset}\t{}",
+                displayname,
+                t.resolve_body(),
+                color_user = color_user,
+                color_reset = color_reset
+            ),
+            Emote(e) => format!(
+                "{color_user}{}{color_reset}\t{}",
```

Comment:
Emotes should be displayed as IRC `/me` messages, those use the `action` prefix.